### PR TITLE
Dashboard header from client cookies - CONSULT

### PR DESCRIPTION
**Pull request by mehalshah**
Rely on the Client Side cookie instead of the session cookie for keeping track of progress in HOC

### CODE REVIEW

File: dashboard/app/assets/javascripts/header.js.erb [hunk 1]
```diff
@@ -25,7 +25,10 @@ if (!window.dashboard) {
  */
 dashboard.buildHeader = function (stageData, progressData, currentLevelId, userId, sectionId) {
   stageData = stageData || {};
-  progressData = progressData || {};
+
+  if (progressData == null) {
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: s/==/===

File: dashboard/app/assets/javascripts/header.js.erb [hunk 2]
```diff
@@ -413,3 +416,46 @@ dashboard.header.updateTimestamp = function () {
     $('.project_updated_at').text("Not saved"); // TODO i18n
   }
 };
+
+dashboard.header.getSummarizedProgressForAnonymousUser = function() {
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: It looks like this function is only used locally. As such, it probably makes sense to not have it live on the dashboard.header namespace, but instead just be function getSummarizedProgressForAnonymousUser.

Really, we should also wrap this entire thing in an anonymous function so that local methods like remixProject don't end up in the global namespace, and only functions explicitly declared on our namespace are "exported". Probably out of scope for this PR though.

Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: nit: space between function and ()

File: dashboard/app/assets/javascripts/header.js.erb [hunk 3]
```diff
@@ -413,3 +416,46 @@ dashboard.header.updateTimestamp = function () {
...
+  var summarizedProgress = {};
+  var levelProgress = {};
+  var cookie = document.cookie.split(';')
+
+  for(var i = 0; i < cookie.length; i++) {
+    var key = cookie[i].split('=')[0].trim().toLowerCase();
+
+    switch(key) {
+      case 'progress':
+        var allLevelsProgress = JSON.parse(decodeURIComponent(pair[1]));
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: where does pair come from?

File: dashboard/app/assets/javascripts/header.js.erb [hunk 4]
```diff
@@ -413,3 +416,46 @@ dashboard.header.updateTimestamp = function () {
...
+            } else if (individualLevelProgress >= 30) {
+              individualLevelProgressDescription = 'perfect';
+            } else if (individualLevelProgress >= 20) {
+              individualLevelProgressDescription = 'passed';
+            } else {
+              individualLevelProgressDescription = 'attempted';
+            }
+
+            levelProgress[level] = {};
+            levelProgress[level]['status'] = individualLevelProgressDescription;
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: We prefer the form levelProgress[level].status. In apps and shared we lint for this. Unfortunately, we don't (yet) have linting for dashboard JS.

File: dashboard/app/assets/javascripts/header.js.erb (hunk 1 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/header.js.erb]: Got it

File: dashboard/app/assets/javascripts/header.js.erb (hunk 3 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/header.js.erb]: Whoops - blew up the variable when cleaning up something. Will put it back

File: dashboard/app/assets/javascripts/header.js.erb (hunk 4 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/header.js.erb]: Will do

File: dashboard/app/assets/javascripts/header.js.erb [hunk 5]
```diff
@@ -25,7 +25,10 @@ if (!window.dashboard) {
  */
 dashboard.buildHeader = function (stageData, progressData, currentLevelId, userId, sectionId) {
   stageData = stageData || {};
-  progressData = progressData || {};
+
+  if (progressData === null) {
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/header.js.erb:31]: Please add a comment explaining that progressData is provided for signed users only.

File: dashboard/app/assets/javascripts/header.js.erb [hunk 6]
```diff
@@ -413,3 +416,47 @@ dashboard.header.updateTimestamp = function () {
     $('.project_updated_at').text("Not saved"); // TODO i18n
   }
 };
+
+function getSummarizedProgressForAnonymousUser () {
+  var summarizedProgress = {};
+  var levelProgress = {};
+  var cookie = document.cookie.split(';');
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/header.js.erb]: Please use a well-tested JS cookie handling library (hopefully we have one already) rather than doing inline cookie cracking which can be error-prone.

File: dashboard/app/assets/javascripts/header.js.erb [hunk 7]
```diff
@@ -413,3 +416,47 @@ dashboard.header.updateTimestamp = function () {
...
+  
+  for(var i = 0; i < cookie.length; i++) {
+    var pair = cookie[i].split('=');
+    var key = pair[0].trim();
+
+    switch(key) {
+      case 'progress':
+        var allLevelsProgress = JSON.parse(decodeURIComponent(pair[1]));
+        for(var level in allLevelsProgress) {
+          if (allLevelsProgress.hasOwnProperty(level)) {
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: One nit that's somewhat personal preference, you could reduce nesting here by saying

```
if (!allLevelsProgress.hasOwnProperty(level)) {
  break;
}
...
```

Also, because this is an object we got form JSON, I think the hasOwnProperty check might be superfluous (will always return true).

You could also do this using forEach, which is what I personally tend to do, but again that's largely personal preference and either approach is fine.

```
allLevelsProgress.forEach(function (level) {
 ...
});
```

File: dashboard/app/assets/javascripts/header.js.erb [hunk 8]
```diff
@@ -413,3 +416,47 @@ dashboard.header.updateTimestamp = function () {
...
+
+            levelProgress[level] = {};
+            levelProgress[level].status = individualLevelProgressDescription;
+          }
+        }
+
+        summarizedProgress['levels'] = levelProgress;
+        break;
+      case 'lines':
+        summarizedProgress['lines'] = pair[1];
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: lines 453 and 456 should also use the dot form.

File: dashboard/app/assets/javascripts/header.js.erb (hunk 6 above)
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: I don't know that we do have one, as I don't know that we do much with cookies on the client currently. Or at least I can't think of places where we do.

File: dashboard/app/assets/javascripts/header.js.erb [hunk 9]
```diff
@@ -413,3 +416,47 @@ dashboard.header.updateTimestamp = function () {
...
+
+    switch(key) {
+      case 'progress':
+        var allLevelsProgress = JSON.parse(decodeURIComponent(pair[1]));
+        for(var level in allLevelsProgress) {
+          if (allLevelsProgress.hasOwnProperty(level)) {
+            var individualLevelProgressDescription = '';
+            var individualLevelProgress = allLevelsProgress[level];
+
+            //TODO: Find a way to use the constants defined somewhere
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/header.js.erb]: There are ruby constants for these values so you can assign their values in an erb file; 

e.g  var x = #{Foo.CLASS_CONSTANT}

I would suggest keeping as much of the code as possible free of Ruby interpolation so that we can split out the truly static scripts from those that require Ruby execution.  (This will help with lintng and efficiency.)

File: dashboard/app/assets/javascripts/header.js.erb [hunk 10]
```diff
@@ -413,3 +416,47 @@ dashboard.header.updateTimestamp = function () {
...
+            if (individualLevelProgress >= 1000) {
+              individualLevelProgressDescription = 'submitted';
+            } else if (individualLevelProgress >= 30) {
+              individualLevelProgressDescription = 'perfect';
+            } else if (individualLevelProgress >= 20) {
+              individualLevelProgressDescription = 'passed';
+            } else {
+              individualLevelProgressDescription = 'attempted';
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/header.js.erb]: Although it probably shouldn't happen, you should do something sensible for individualLevelProgress = 0, which is not even attempted.

File: dashboard/app/assets/javascripts/header.js.erb [hunk 11]
```diff
@@ -413,3 +416,47 @@ dashboard.header.updateTimestamp = function () {
...
+  for(var i = 0; i < cookie.length; i++) {
+    var pair = cookie[i].split('=');
+    var key = pair[0].trim();
+
+    switch(key) {
+      case 'progress':
+        var allLevelsProgress = JSON.parse(decodeURIComponent(pair[1]));
+        for(var level in allLevelsProgress) {
+          if (allLevelsProgress.hasOwnProperty(level)) {
+            var individualLevelProgressDescription = '';
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/header.js.erb]: Consider adding a getLevelProgressDescription(progress) helper for readability and testability.

We tend to be a bit more concise in our variable names;  I think just levelProgressDescription is perfectly clear in this context.

File: dashboard/app/assets/javascripts/header.js.erb (hunk 5 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/header.js.erb:31]: Added

File: dashboard/app/views/layouts/_header.html.haml [hunk 12]
```diff
@@ -55,5 +55,5 @@
 - if script_level
   :javascript
     //<![CDATA[
-    dashboard.buildHeader(#{script_level.stage.summarize.to_json}, #{summarize_user_progress(script_level.script, @user || current_user).to_json}, #{script_level.level_id}, #{@user.try(:id) || 'null'}, #{@section.try(:id) || 'null'});
+    dashboard.buildHeader(#{script_level.stage.summarize.to_json}, #{current_user == nil ? 'null' : summarize_user_progress(script_level.script, @user || current_user).to_json}, #{script_level.level_id}, #{@user.try(:id) || 'null'}, #{@section.try(:id) || 'null'});
```
Comment (philbogle) [inline dashboard/app/views/layouts/_header.html.haml]: The ternary expression is a bit too complex to include inline, probably better to assign the value to a variable if the ruby code after the if statement.

File: dashboard/app/assets/javascripts/header.js.erb (hunk 6 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/header.js.erb]: I spent a little time looking at options and didn't use one because of Brent's reason. If we're going to do a lot more handling then I'll find something appropriate.

File: dashboard/app/assets/javascripts/header.js.erb (hunk 8 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/header.js.erb]: Will change

File: dashboard/app/assets/javascripts/header.js.erb (hunk 10 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/header.js.erb]: Will do

File: dashboard/app/assets/javascripts/header.js.erb (hunk 6 above)
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: I think there's probably value in finding a good library to do cookie parsing now that we can use going forward. I don't personally have a lot of experience with cookie libraries. One that Phil referenced/found on google that looks both small and promising is https://github.com/ScottHamper/Cookies

File: dashboard/app/assets/javascripts/header.js.erb (hunk 7 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/header.js.erb]: If the check is superfluous then I'll just remove it

File: dashboard/app/views/layouts/_header.html.haml (hunk 12 above)
Comment (mehalshah) [inline dashboard/app/views/layouts/_header.html.haml]: I'll rewrite it

File: dashboard/app/assets/javascripts/header.js.erb [hunk 13]
```diff
@@ -413,3 +419,35 @@ dashboard.header.updateTimestamp = function () {
     $('.project_updated_at').text("Not saved"); // TODO i18n
   }
 };
+
+function getSummarizedProgressForAnonymousUser () {
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/header.js.erb:427]: Please document this method, including jsdoc comments describing the return value.

File: dashboard/app/assets/javascripts/header.js.erb [hunk 14]
```diff
@@ -413,3 +419,35 @@ dashboard.header.updateTimestamp = function () {
...
+  var summarizedProgress = {};
+  var levelProgress = {};
+
+  summarizedProgress.lines = $.cookie('lines');
+
+  var allLevelsProgress = JSON.parse(decodeURIComponent($.cookie('progress')));
+  for(var level in allLevelsProgress) {
+    var individualLevelProgressDescription = '';
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/header.js.erb]: Replace individualLevelProgress with levelProgress (and likewise for the description version)

individualLevelProgressDescription is so long I find it hurts rather than helps readability.  Also, I believe we don't currently compile or minify these names, so it bloats the size of the file.  (This is a lesser concern but somewhat relevant.)

File: dashboard/lib/client_state.rb [hunk 15]
```diff
@@ -13,9 +13,8 @@ def initialize(session, cookies)
 
   # Resets all client state (level progress, lines of code, videos seen, etc.)
   def reset
-    # Also reset unencrypted cookies in case we are in a rolled-back state.
-    session[:lines] = cookies[:lines] = nil
-    session[:progress] = cookies[:progress] = nil
+    cookies[:lines] = nil
```
Comment (philbogle) [inline dashboard/lib/client_state.rb:16]: These changes are now checked in; could you pull from staging so we don't see them?

**Comment by philbogle** [comment]
As discussed, please add eye tests that verify that the client correctly shows progress after completing a level.  This will let us be confident these changes work across all browsers.

**Comment by Bjvanminnen** [comment]
As a point of clarity, eyes tests run only on chrome. Other non eyes ui
tests hit multiple browsers

**Comment by philbogle** [comment]
Ah, what is the best path to make sure we have cross-browser coverage?

**Comment by Bjvanminnen** [comment]
My point was mainly one of terminology. UI tests refers to all of the
.feature files we have in dashboard/tests/ui. These are driven by
cucumber/selenium. Most of them run against all browsers.

A subset of those .feature files have scenarios marked with @eyes. These
are our eyes tests. They are also driven by cucumber/selenium, but differ
in that at various points they take screenshots (and compare said
screenshots to baselines from previous runs). Right now, they only run on
Chrome. It's possible that we could/should expand coverage, though these
are fairly noisy and expanding coverage would have a cost.

tldr version = eyes tests are a subset of UI tests that take screenshots
and only run on chrome (right now)

File: dashboard/app/assets/javascripts/header.js.erb (hunk 13 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/header.js.erb:427]: Done

File: dashboard/app/assets/javascripts/header.js.erb (hunk 14 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/header.js.erb]: I'm of the belief that verbose names are generally better as they remove the need for comments describing what a variable is. I can still make the change to fix this though.

If we're worried about file size, then we absolutely should be minifying names. Let's put that in the backlog.

Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js.erb]: These do appear to get minified on production - they end up in application-<digest>.js. Names on the dashboard namespace will not be minified (i.e. you can search for dashboard.buildHeader), but local function names will be (i.e. you won't find a showHeaderPopup).

File: dashboard/test/ui/features/hourOfCode.feature [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+  And I close the dialog
+  Then I wait until element "#runButton" is visible
+  And I press "runButton"
+  Then I wait to see ".modal"
+  Then I close the dialog
+  Then I reload the page
+  Then I wait to see ".modal"
+  And I close the dialog
+  When element "#runButton" is visible
+  Then element ".header_middle a:first" has class "level_link attempted"gi
```
Comment (Bjvanminnen) [inline dashboard/test/ui/features/hourOfCode.feature]: trailing gi?

**Comment by Bjvanminnen** [comment]
lgtm
